caddytls/certificate_transparency.go

```diff
@@ -0,0 +1,123 @@
+package caddytls
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// This code borrows heavily from https://github.com/grahamedgecombe/ct-submit
+
+type addChainRequest struct {
+	Chain []string `json:"chain"`
+}
+
+type signedCertificateTimestamp struct {
+	Version    uint8  `json:"sct_version"`
+	LogID      string `json:"id"`
+	Timestamp  int64  `json:"timestamp"`
+	Extensions string `json:"extensions"`
+	Signature  string `json:"signature"`
+}
+
+// AsRawBytes converts the SCT from the JSON form that the submission API
+// returns into the raw binary form that the TLS handshake expects.
+func (sct *signedCertificateTimestamp) AsRawBytes() ([]byte, error) {
+	b := new(bytes.Buffer)
+	// Version
+	binary.Write(b, binary.BigEndian, sct.Version)
+
+	// LogID
+	logBytes, err := base64.StdEncoding.DecodeString(sct.LogID)
+	if err != nil {
+		return nil, err
+	}
+	b.Write(logBytes)
+
+	// Timestamp
+	binary.Write(b, binary.BigEndian, sct.Timestamp)
+
+	// Extensions
+	extensionBytes, err := base64.StdEncoding.DecodeString(sct.Extensions)
+	if err != nil {
+		return nil, err
+	}
+	if len(extensionBytes) > 65535 {
+		return nil, fmt.Errorf("Extensions is too long: %d bytes", len(extensionBytes))
+	}
+	binary.Write(b, binary.BigEndian, uint16(len(extensionBytes)))
+	b.Write(extensionBytes)
+
+	// Signature
+	signatureBytes, err := base64.StdEncoding.DecodeString(sct.Signature)
+	if err != nil {
+		return nil, err
+	}
+	b.Write(signatureBytes)
+
+	return b.Bytes(), nil
+}
+
+// Makes an HTTP request to the log server and returns the parsed SCT response.
+// If the cert is already in the log, the log will simply return the previously
+// generated SCT, making this idempotent.
+func submitSCT(client http.Client, url string, payload []byte) (*signedCertificateTimestamp, error) {
+	if !strings.Contains(url, "://") {
+		// Allow people to omit the protocol and default to HTTPS.
+		url = "https://" + url
```

Comment:
Certificate Transparency requires the use of HTTPS, so I suggest always prepending `https://` and requiring logs to be specified without the protocol, as this is how log addresses are conventionally represented.